src/library.cpp

```diff
@@ -157,6 +157,127 @@ bool Library::removeBookmark(const std::string& zimId, const std::string& url)
   return false;
 }
 
+std::tuple<int, int> Library::migrateBookmarks() {
+  std::set<std::string> sourceBooks;
+  int invalidBookmarks = 0;
+  {
+    std::lock_guard<std::recursive_mutex> lock(m_mutex);
+    for(auto& bookmark:m_bookmarks) {
+      if (m_books.find(bookmark.getBookId()) == m_books.end()) {
+        invalidBookmarks += 1;
+        sourceBooks.insert(bookmark.getBookId());
+      }
+    }
+  }
+  int changed = 0;
+  for(auto& sourceBook:sourceBooks) {
+    changed += migrateBookmarks(sourceBook, ALLOW_DOWNGRADE);
+  }
+  return std::make_tuple(changed, invalidBookmarks);
+}
+
+void Library::removeOlderBooksFromBookCollection(BookIdCollection& books, const Bookmark& bookmark) const {
+  auto foundIt = std::find(books.begin(), books.end(), bookmark.getBookId());
+  if (foundIt == books.end()) {
+    foundIt = std::find_if(books.begin(), books.end(), [&](const std::string& bookId) {
+      auto& book = getBookById(bookId);
+      return book.getDate() <= bookmark.getDate();
+    });
+  }
+  if (foundIt != books.end()) {
+    // books is sorted by date, remove the source and all older books
+    books.erase(foundIt, books.end());
+  }
+}
+
+std::string remove_quote(std::string input) {
+  std::replace(input.begin(), input.end(), '"', ' ');
+  return input;
+}
+
+std::string Library::getBestTargetBookId(const Bookmark& bookmark, MigrationMode migrationMode) const {
+  std::lock_guard<std::recursive_mutex> lock(m_mutex);
+  // Search for a existing book with the same name
+  auto book_filter = Filter();
+  if (!bookmark.getBookName().empty()) {
+    book_filter.name(bookmark.getBookName());
+  } else {
+    // We don't have a name stored (older bookmarks)
+    // Fallback on title (All bookmarks should have one, but let's be safe against wrongly filled bookmark)
+    if (bookmark.getBookTitle().empty()) {
+        // No bookName nor bookTitle, no way to find target book.
+        return "";
+    }
+    book_filter.query("title:\"" + remove_quote(bookmark.getBookTitle()) + "\"");
+  }
+  auto targetBooks = filter(book_filter);
+  sort(targetBooks, DATE, false);
+  if (migrationMode == UPGRADE_ONLY) {
+    removeOlderBooksFromBookCollection(targetBooks, bookmark);
+  }
+    
+  if (targetBooks.empty()) {
+    // No better book to migrate to.
+    // If bookmark is valid, best is the current one.
+    try {
+      getBookById(bookmark.getBookId());
+      return bookmark.getBookId();
+    } catch (std::out_of_range& e) {
+      return "";
+    }
+  }
+  if (targetBooks.size() != 1) {
+    // We have several, reduce to same flavour
+    auto flavouredTargetBooks = filter(book_filter.flavour(bookmark.getBookFlavour()));
+    sort(flavouredTargetBooks, DATE, false);
+    if (migrationMode == UPGRADE_ONLY) {
+      removeOlderBooksFromBookCollection(flavouredTargetBooks, bookmark);
+    }
+    if (!flavouredTargetBooks.empty()) {
+      targetBooks = flavouredTargetBooks;
+    }
+  }
```

Comment:
I think that the code duplication in this function introduced by following my previous suggestion as well as my other comment on the newer-different-flavour-vs-older-same-flavour situation can be resolved as follows.

It doesn't make sense running the filter query a second time. After it turns out that there is more than a single candidate according to the simple criteria we can add more criteria and filter or reorder the result further. Thus the function `pruneBookCollectionFromOlderBooks()` from the previous version of this PR can be renamed to `getPrioritizedMigrationCandidates()` (possibly returning the result instead of modifying the input collection) and extended with a step that runs an additional sort on the remaining books. I think that such a design will allow for easier addition of extra criteria for selecting the best target book.

Reply:
Nice input.
I have rewrite the algorithm with your idea.